lib/src/modules/task.cc

```diff
@@ -128,11 +136,171 @@ static TaskCommand getTaskCommand(const fs::path &task_executable)
     return TaskCommand { task_executable.string(), { } };
 }
 
+// Converts boost filesystem errors to a task_error object.
+// TODO: Modify this to convert it to a PXP-error instead.
+// All download code should be in a try-catch block (just
+// one).
+static Module::ProcessingError toModuleProcessingError(const fs::filesystem_error& e) {
+    std::string err_msg = e.what();
+    if (alg::contains(err_msg, "No such file or directory")) {
+      return Module::ProcessingError(lth_loc::format("No such file or directory: {1}", e.path1()));
+    } else if (alg::contains(err_msg, "File exists")) {
+      return Module::ProcessingError(lth_loc::format("A file matching the name of the provided sha already exists"));
+    }
+    return Module::ProcessingError(err_msg);
+}
+
+// Take out common permissions.
+static fs::perms TASK_FILE_RW_PERMS = fs::owner_read | fs::owner_write | fs::group_read | fs::group_write;
+
+// Creates the <task_cache_dir>/<sha256> directory for a single
+// task and ensures that its permissions are only readable/writable by
+// the PXP agent owner/group. Returns the path to this directory.
+// Note that the last modified time of the directory is updated.
+static fs::path createCacheDir(const fs::path& task_cache_dir, const std::string& sha256) {
+    auto cache_dir = fs::path(task_cache_dir) / sha256;
+    fs::create_directory(cache_dir);
+    fs::last_write_time(cache_dir, time(nullptr));
+    fs::permissions(cache_dir, TASK_FILE_RW_PERMS);
```

Comment:
https://github.com/puppetlabs/pxp-agent/pull/623 has some examples for spool-dir and logfile.